src/core/mempool.rs

```diff
@@ -400,244 +391,85 @@ impl MemPoolDB {
     where
         F: FnMut(Vec<MemPoolTxInfo>) -> Result<(), E>,
         E: From<db_error> + From<ChainstateError>,
     {
-        let (mut tip_consensus_hash, mut tip_block_hash, mut tip_height) = (
-            tip_consensus_hash.clone(),
-            tip_block_hash.clone(),
-            tip_height,
-        );
-
-        debug!(
-            "Begin scanning transaction mempool at {}/{} height={}",
-            &tip_consensus_hash, &tip_block_hash, tip_height
-        );
+        // Want to consider transactions with
+        // height > max(0, `tip_height - MEMPOOL_MAX_TRANSACTION_AGE`) - 1
+        let min_height = match tip_height.checked_sub(MEMPOOL_MAX_TRANSACTION_AGE) {
+            None => None,
+            Some(h) if h == 0 => None,
+            Some(h) => Some(h - 1),
+        };
+        let mut curr_page = 0;
 
-        let mut next_timestamp =
-            match MemPoolDB::get_next_timestamp(&self.db, &tip_consensus_hash, &tip_block_hash, 0)?
-            {
-                Some(ts) => ts,
-                None => loop {
-                    // walk back to where the first transaction we can mine can be found
-                    match self.walk(chainstate, &tip_consensus_hash, &tip_block_hash, tip_height)? {
-                        MemPoolWalkResult::Chainstate(
-                            next_consensus_hash,
-                            next_block_bhh,
-                            next_height,
-                            next_timestamp,
-                        ) => {
-                            tip_consensus_hash = next_consensus_hash;
-                            tip_block_hash = next_block_bhh;
-                            tip_height = next_height;
-                            break next_timestamp;
-                        }
-                        MemPoolWalkResult::NoneAtHeight(
-                            next_consensus_hash,
-                            next_block_hash,
-                            next_height,
-                        ) => {
-                            if std::env::var("STACKS_MEMPOOL_BAD_BEHAVIOR") == Ok("1".into()) {
-                                warn!(
-                                "Stopping mempool walk because no mempool entries at height = {}",
-                                next_height - 1
-                            );
-                                return Ok(());
-                            } else {
-                                tip_consensus_hash = next_consensus_hash;
-                                tip_block_hash = next_block_hash;
-                                tip_height = next_height;
-                            }
-                        }
-                        MemPoolWalkResult::Done => {
-                            return Ok(());
-                        }
-                    }
-                },
-            };
+        let (mut next_nonce, mut next_timestamp) = match MemPoolDB::get_next_nonce_and_timestamp(
```

Comment:
My understanding of why we previously were using the timestamp in iteration was as a way to do some paging -- with the new structure of the mempool with explicit `cur_page` parameters, I don't think we need to use the timestamp anymore, instead we could just iterate using the nonce and the cur_page parameter (still including the timestamp in the sort though). This would simplify the loop and also reduce the number of SQLite queries.

Reply:
Oh ok, I thought we were using the timestamp for a different reason. I'll simplify the logic here then.